lookout/style/typos/preprocessing/create_typos.py

```diff
@@ -0,0 +1,155 @@
+from os import path
+import random
+import string
+from typing import Tuple
+
+import pandas
+from tqdm import tqdm
+
+from lookout.style.typos.utils import COLUMNS
+
+letters = list(string.ascii_lowercase)
+
+
+def rand_bool(true_prob: float):
+    """
+    Return True with probability true_prob.
+    """
+    return random.uniform(0, 1) < true_prob
+
+
+def rand_insert(token: str):
+    """
+    Add random letter inside a token.
+    """
+    letter = random.choice(letters)
+    if len(token) == 0:
+        return letter
+
+    pos = random.choice(range(len(token) + 1))
+    if pos == len(token):
+        return token + letter
+    return token[:pos] + letter + token[pos:]
+
+
+def rand_delete(token: str):
+    """
+    Delete random symbol from a token.
+    """
+    if len(token) == 0:
+        return token
+    pos = random.choice(range(len(token)))
+    return token[:pos] + token[pos + 1:]
+
+
+def rand_substitution(token: str):
+    """
+    Substitute random symbol with a letter inside a token.
+    """
+    if len(token) == 0:
+        return token
+    pos = random.choice(range(len(token)))
+    letter = random.choice(letters)
+    return token[:pos] + letter + token[pos + 1:]
+
+
+def rand_swap(token: str):
+    """
+    Swap two random consequent symbols.
+    """
+    if len(token) < 2:
+        return token
+    pos = random.choice(range(len(token) - 1))
+    return token[:pos] + token[pos + 1] + token[pos] + token[pos + 2:]
+
+
+def rand_typo(token: str):
+    """
+    Make random typo in a token.
+    """
+    typo_func = random.choice([rand_insert, rand_delete, rand_substitution, rand_swap])
+    return typo_func(token)
+
+
+def corrupt_tokens_in_df(data: pandas.DataFrame, typo_probability: float,
+                         add_typo_probability: float) -> pandas.DataFrame:
+    """
+    Create artificial typos in tokens from a dataframe.
+
+    Augment some of identifiers from dataframe with TYPO_PROBABILITY,
+    consequent typos in the same word happen with ADD_TYPO_PROBABILITY each.
+    Operations happens inplace.
+    """
+    typoed_tokens = list(data[COLUMNS["TOKEN"]])
+    if COLUMNS["SPLIT"] in data.columns:
+        typoed_token_split = list(data[COLUMNS["SPLIT"]])
+
+    for row_number in tqdm(range(len(data))):
```

Comment:
We need to add `log_level` argument and do the same as in https://github.com/src-d/style-analyzer/blob/master/lookout/style/format/rules.py#L883